src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Automata.cs

```diff
@@ -0,0 +1,429 @@
+
+        /// <summary>Gets <see cref="_nfaStateArrayInverse"/>.Count</summary>
+        private int NfaStateCount => _nfaStateArrayInverse.Count;
+
+        /// <summary>
+        /// Transition function for NFA transitions in NFA mode.
+        /// Each NFA entry maps to a list of NFA target states.
+        /// Each list of target states is without repetitions.
+        /// If the entry is null then the targets states have not been computed yet.
+        /// </summary>
+        private int[]?[] _nfaDelta = Array.Empty<int[]>();
+
+        /// <summary>
+        /// The transition function for <see cref="FindSubcaptures(ReadOnlySpan{char}, int, int, PerThreadData)"/>,
+        /// which is an NFA mode with additional state to track capture start and end positions.
+        /// Each entry is an array of pairs of target state and effects to be applied when taking the transition.
+        /// If the entry is null then the transition has not been computed yet.
+        /// </summary>
+        private (int, DerivativeEffect[])[]?[] _capturingNfaDelta = Array.Empty<(int, DerivativeEffect[])[]?>();
+
+        /// <summary>
+        /// Implements a version of <see cref="Array.Resize"/> that is guaranteed to not publish an array before values
+        /// have been copied over.
+        /// </summary>
+        /// <remarks>
+        /// This may not be strictly necessary for arrays of primitive or reference types (which have atomic
+        /// reads/writes), as when, e.g., <see cref="_delta"/> is found to not have an entry the array is checked again
+        /// after a lock on the matcher has been acquired. However, in a highly threaded use case it still seems better
+        /// to avoid unnecessarily causing other threads to acquire the lock.
+        /// </remarks>
+        private static void ConcurrentArrayResize<T>(ref T[] array, int newSize)
+        {
+            Debug.Assert(newSize >= array.Length);
+            T[] newArray = new T[newSize];
+            Array.Copy(array, newArray, array.Length);
+            Volatile.Write(ref array, newArray);
+        }
+
+        private int DeltaOffset(int stateId, int mintermId) => (stateId << _mintermsLog) | mintermId;
+
+        /// <summary>Returns the span from <see cref="_delta"/> that may contain transitions for the given state</summary>
+        private Span<int> GetDeltasFor(DfaMatchingState<TSet> state)
+        {
+            int numMinterms = _minterms.Length;
+            if (state.StartsWithLineAnchor)
+            {
+                numMinterms++;
+            }
+
+            return _delta.AsSpan(state.Id << _mintermsLog, numMinterms);
+        }
+
+        /// <summary>Returns the span from <see cref="_nfaDelta"/> that may contain transitions for the given state</summary>
+        private Span<int[]?> GetNfaDeltasFor(DfaMatchingState<TSet> state)
+        {
+            Debug.Assert(Monitor.IsEntered(this));
```

Comment:
It's guaranteed to be held for both in current usage, but didn't seem strictly required for GetDeltasFor. I can add it so that if that guarantee ever changes someone will have to take a closer look :)